jib-core/src/main/java/com/google/cloud/tools/jib/ncache/CacheWriteEntry.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.ncache;
+
+import com.google.cloud.tools.jib.blob.Blob;
+import com.google.cloud.tools.jib.image.DescriptorDigest;
+import java.util.Optional;
+
+/** Represents layer data to write to the cache. <b>Implementations must be immutable.</b> */
+public interface CacheWriteEntry {
+
+  /**
+   * Gets the {@link Blob} to write as the layer contents.
+   *
+   * @return the layer {@link Blob}
+   */
+  Blob getLayerBlob();
+
+  /**
+   * Gets the optional selector digest to also reference this layer data.
+   *
+   * @return the selector digest
+   */
+  Optional<DescriptorDigest> getSelector();
+
+  /**
+   * Gets the optional {@link Blob} to write as the arbitrary layer metadata.
+   *
+   * @return the metadata {@link Blob}
+   */
+  Optional<Blob> getMetadataBlob();
```

Comment:
Same question as "selector" as to who would and wouldn't provide a value for this.

Reply:
Same as selector - application layers will have metadata, base image layers will not.

Reply:
also here, perhaps some more detail.

```
.... to write as the arbitrary layer metadata (for example: last modified time, file hashes, etc)
```